store/src/java/com/zimbra/cs/event/Event.java

```diff
@@ -157,19 +165,38 @@ public static Event generateSentEvent(String accountId, int messageId, String se
     /**
      * Convenience method to generate a SENT event for every recipient of the email
      */
-    public static List<Event> generateSentEvents(String accountId, int messageId, Address sender, Address[] allRecipients, String dsId, Long timestamp) {
-        List<Event> sentEvents = new ArrayList<>(allRecipients.length);
-        for (Address address : allRecipients) {
-            sentEvents.add(generateSentEvent(accountId, messageId, sender.toString(), address.toString(), dsId, timestamp));
+    public static List<Event> generateSentEvents(Message msg, Long timestamp) {
+        try {
+            String acctId = msg.getAccountId();
+            int msgId = msg.getId();
+            String dsId = msg.getDataSourceId();
+            ParsedMessage pm = msg.getParsedMessage();
+            MimeMessage mm = pm.getMimeMessage();
+            Address[] recipients = mm.getAllRecipients();
+            Address sender = mm.getFrom()[0];
```

Comment:
Need to handle mm.getFrom() returning null:

From https://docs.oracle.com/javaee/7/api/javax/mail/internet/MimeMessage.html#getFrom--

    public Address[] getFrom()
                  throws MessagingException

    Returns the value of the RFC 822 "From" header fields. If this header field is absent, the "Sender" header field is used. If the "Sender" header field is also absent, null is returned.

Not sure about getAllRecipients() return - probably best to handle null return for that too.
May want to include test data to exercise this.

Reply:
Ah yes, good point.

Reply:
Actually, probably event better to use `ParsedMessage.getSenderEmail()`